src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1064 @@
+  if (data_indices != nullptr && num_data != num_data_) {
+    indices_future_ = boost::compute::copy_async(data_indices, data_indices + num_data, device_data_indices_->begin(), queue_);
+  }
+  // generate and copy ordered_gradients if gradients is not null
+  if (gradients != nullptr) {
+    if (num_data != num_data_) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data; ++i) {
+        ordered_gradients[i] = gradients[data_indices[i]];
+      }
+      gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, num_data * sizeof(score_t), ptr_pinned_gradients_);
+    }
+    else {
+      gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, num_data * sizeof(score_t), gradients);
+    }
+  }
+  // generate and copy ordered_hessians if hessians is not null
+  if (hessians != nullptr) {
+    if (num_data != num_data_) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data; ++i) {
+        ordered_hessians[i] = hessians[data_indices[i]];
+      }
+      hessians_future_ = queue_.enqueue_write_buffer_async(device_hessians_, 0, num_data * sizeof(score_t), ptr_pinned_hessians_);
+    }
+    else {
+      hessians_future_ = queue_.enqueue_write_buffer_async(device_hessians_, 0, num_data * sizeof(score_t), hessians);
+    }
+  }
+  // converted indices in is_feature_used to feature-group indices
+  std::vector<int8_t> is_feature_group_used(num_feature_groups_, 0);
+  for (int i = 0; i < num_features_; ++i) {
+    if(is_feature_used[i]) {
+      is_feature_group_used[train_data_->Feature2Group(i)] = 1;
+    }
+  }
+  // construct the feature masks for dense feature-groups
+  int used_dense_feature_groups = 0;
+  for (int i = 0; i < num_dense_feature_groups_; ++i) {
+    if (is_feature_group_used[dense_feature_group_map_[i]]) {
+      feature_masks_[i] = 1;
+      ++used_dense_feature_groups;
+    }
+    else {
+      feature_masks_[i] = 0;
+    }
+  }
+  bool use_all_features = used_dense_feature_groups == num_dense_feature_groups_;
+#if GPU_DEBUG >= 1
+  printf("feature masks:\n");
+  for (unsigned int i = 0; i < feature_masks_.size(); ++i) {
+    printf("%d ", feature_masks_[i]);
+  }
+  printf("\n");
+  printf("%d feature groups, %d used, %d\n", num_dense_feature_groups_, used_dense_feature_groups, use_all_features);
+#endif
+  if (!use_all_features) {
+    queue_.enqueue_write_buffer(device_feature_masks_, 0, num_dense_feature4_ * dword_features_, ptr_pinned_feature_masks_);
+  }
+  // All data have been prepared, now run the GPU kernel
+  GPUHistogram(num_data, use_all_features);
+  return true;
+}
+
+void GPUTreeLearner::ConstructHistograms(const std::vector<int8_t>& is_feature_used, bool use_subtract) {
+  std::vector<int8_t> is_sparse_feature_used(num_features_, 0);
+  std::vector<int8_t> is_dense_feature_used(num_features_, 0);
+  for (int feature_index = 0; feature_index < num_features_; ++feature_index) {
```

Comment:
change all loop with ```num_features_``` and ```num_group``` (like https://github.com/Microsoft/LightGBM/blob/master/src/treelearner/serial_tree_learner.cpp#L450-L459 and ) to multi-threading? Or it will be slow when #feature is large.